types/react/index.d.ts

```diff
@@ -2633,30 +2658,76 @@ declare namespace React {
-    ? Defaultize<MergePropTypes<P, PropTypes.InferProps<T>>, D>
-    : C extends { propTypes: infer T; }
-        ? MergePropTypes<P, PropTypes.InferProps<T>>
-        : C extends { defaultProps: infer D; }
-            ? Defaultize<P, D>
-            : P;
+    const Invalid_propTypes: unique symbol;
+    /**
+     * Dummy interface that is intersected into the props type of a component on
+     * use to generate an error if the component's `propTypes` static property
+     * is invalid.  In most cases, you can get more information about the
+     * problem with the `propTypes` by trying to assign the component to a
+     * variable of type `React.ComponentType<P>` where `P` is the appropriate
+     * props type.  You can bypass the type checking of `propTypes` by defining
+     * a static property named `bypassPropTypesTypecheck`.
+     */
+    interface Invalid_propTypes<P, T> {
+        [Invalid_propTypes]: undefined;
+    }
+
+    const Invalid_defaultProps: unique symbol;
+    /**
+     * Dummy interface that is intersected into the props type of a component on
+     * use to generate an error if the component's `defaultProps` static
+     * property is invalid.  In most cases, you can get more information about
+     * the problem with the `defaultProps` by trying to assign the component to
+     * a variable of type `React.ComponentType<P>` where `P` is the appropriate
+     * props type.  You can bypass the type checking of `defaultProps` by
+     * defining a static property named `bypassDefaultPropsTypecheck`.
+     */
+    interface Invalid_defaultProps<P, D> {
+        [Invalid_defaultProps]: undefined;
+    }
+
+    // For a component class, the P passed to JSX.LibraryManagedAttributes is
+    // based on the `props` property and includes `children` regardless of
+    // whether the original props type (the `P` type argument to
+    // React.Component) does.  However, it's common for component classes not
+    // to list `children` in the `propTypes` if the original props type did not
+    // include `children`.  We accommodate that here by making the validator for
+    // `children` optional.
+    type ValidationMapWithOptionalChildren<P> =
+        ValidationMap<Pick<P, Exclude<keyof P, "children">>> &
+        Partial<ValidationMap<Pick<P, Extract<keyof P, "children">>>>;
+
+    // Explicitly check for excess properties in CheckPropTypes and
+    // CheckDefaultProps because by the time we get to
+    // JSX.LibraryManagedAttributes, we have lost the opportunity to do an
+    // excess properties check on the original object literal.
+
+    type CheckPropTypes<P, T> =
+        // This condition attempts to single out `T = any`, which needs to bypass the rest of the check
+        // because `keyof T` would come out as not extending `keyof P`.
+        T extends typeof Invalid_propTypes ? {} :
+        [keyof T, T] extends [keyof P, ValidationMapWithOptionalChildren<P>] ? {} : Invalid_propTypes<P, T>;
```

Comment:
Neat, I hadn't checked you could use tuple types as a sort of `&&` to avoid nesting